### PR TITLE
add datadog.parse_record_headers property

### PR DESCRIPTION
**Pull request by jm1221d**
### What does this PR do?

Add a property `datadog.parse_record_headers` to the connector configuration that enables parsing the Kafka Record Headers and sending them to DataDog. The attribute would be called `kafkaheaders`.

### Motivation

Currently, Kafka Record Headers are not sent to DataDog. OpenTelemetry injects a record header, called traceparent, to all kafka records. This change would enable us to correlate all records that we received in DataDog by the traceparent record header key or any other record header key. 

### Additional Notes

Additionally, we could add another property that would explicitly contain the header keys that the user wants to parse and send to DataDog.

### CODE REVIEW

File: src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java [hunk 1]
```diff
@@ -97,20 +117,29 @@ private JsonArray formatBatch(String topic) {
...
         return batchRecords;
     }
 
+    private JsonElement kafkaHeadersToJsonElement(SinkRecord sinkRecord) {
+        Map<String, Object> headerMap = stream(sinkRecord.headers().spliterator(), false)
+                .collect(toMap(Header::key, Header::value));
+
+        String jsonString = gson.toJson(headerMap);
+
+        return gson.fromJson(jsonString, JsonElement.class);
```
Comment (remeh) [inline src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java:133]: Could this map possibly contains a lot of keys and/or do we know about all of them before-hand or are they arbitrary?
I'm basically wondering if this `toJson/fromJson` couple could be turned in a more precise serialisation, directly creating a `JsonElement` instead.

File: src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java [hunk 2]
```diff
@@ -97,20 +117,29 @@ private JsonArray formatBatch(String topic) {
...
+    }
+
     private JsonElement recordToJSON(SinkRecord record) {
         byte[] rawJSONPayload = jsonConverter.fromConnectData(record.topic(), record.valueSchema(), record.value());
         String jsonPayload = new String(rawJSONPayload, StandardCharsets.UTF_8);
-        return new Gson().fromJson(jsonPayload, JsonElement.class);
+        return gson.fromJson(jsonPayload, JsonElement.class);
```
Comment (remeh) [inline src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java:139]: Are we confident `gson` is completely stateless? It might looks like a detail, but if we're not 100% sure of this, let's maybe keep this unrelated fix for a separate PR.

File: src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java [hunk 3]
```diff
@@ -97,20 +117,29 @@ private JsonArray formatBatch(String topic) {
             }
 
             JsonElement recordJSON = recordToJSON(record);
-            JsonObject message = populateMetadata(topic, recordJSON, record.timestamp());
+            JsonObject message = populateMetadata(topic, recordJSON, record.timestamp(), () -> kafkaHeadersToJsonElement(record));
```
Comment (remeh) [inline src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java:120]: Shouldn't we have this change behind a `config.parseRecordHeaders` test?
I'm thinking about the extra toJson/fromJson call even for people that don't want to use these information.

**Review by remeh — COMMENTED** [review]
Hey @jm1221d, thanks for the contribution!

I left a few comments and I also have a separate question on the OpenTelemetry implementation you're referring to: is this implementation also adding these headers in a JSON or are they sent differently?

**Comment by jm1221d** [comment]
Hi @remeh , thanks for looking at the PR!

Currently, the OpenTelemetry implementation adds a header called traceparent (or other headers if enabled -> b3), and the value format is described here: https://www.w3.org/TR/trace-context/

File: src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java (hunk 1 above)
Comment (jm1221d) [reply src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java:133]: Headers are arbitrary and are created by the kafka producer. Some frameworks (for example Spring Kafka) have default headers that are added automatically to the kafka record. Headers are key-value pairs (string-bite[]) on the producer side, so a record could potentially have a lot of keys. We could add an additional property that would contain an explicit list of headers that the user would like to inject into DataDog.

I think that we can turn it into a more precise serialisation, will push it to the branch

File: src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java (hunk 2 above)
Comment (jm1221d) [reply src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java:139]: It looks like Gson should be completely stateless, but I saw some issues with it not long ago. Let's keep it the old way then!

File: src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java (hunk 3 above)
Comment (jm1221d) [reply src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java:120]: I've added a test for the `config.parseRecordHeaders`: https://github.com/DataDog/datadog-kafka-connect-logs/pull/49/files#diff-6cefcd933c4ff34173ccb8e393231ab406748e7d05f8f1d3c29221f66ce551ebR201-R202

Also, I've added the parameter as a supplier so that `kafkaHeadersToJsonElement` is not executed if `config.parseRecordHeaders` is not  set to `true`

File: README.md [hunk 4]
```diff
@@ -24,25 +24,25 @@ See [Confluent's documentation](https://docs.confluent.io/current/connect/managi
...
+Also see [Confluent's documentation](https://docs.confluent.io/current/connect/managing/community.html) on installing
 community connectors.
 
 ### Build from Source
 
 1. Clone the repo from https://github.com/DataDog/datadog-kafka-connect-logs
 2. Verify that Java8 JRE or JDK is installed.
-3. Run `mvn clean compile package`. This will build the jar in the `/target` directory. The name will be 
-`datadog-kafka-connect-logs-[VERSION].jar`.
+3. Run `mvn clean compile package`. This will build the jar in the `/target` directory. The name will be
```
Comment (urseberry) [inline README.md:34]:
```suggestion
3. Run `mvn clean compile package`. This builds the jar in the `/target` directory. The file name has the format
```

File: README.md [hunk 5]
```diff
@@ -24,25 +24,25 @@ See [Confluent's documentation](https://docs.confluent.io/current/connect/managi
...
+   `datadog-kafka-connect-logs-[VERSION].jar`.
 4. The zip file for use on [Confluent Hub](https://www.confluent.io/hub/) can be found in `target/components/packages`.
 
 ## Quick Start
 
 1. To install the plugin, place the plugin's jar file (see [previous section](#installation-and-setup) on how to download or build it)
-in or under the location specified in `plugin.path` . If you use Confluent Platform, simply run 
-`confluent-hub install target/components/packages/<connector-zip-file>`.
+   in or under the location specified in `plugin.path` . If you use Confluent Platform, simply run
```
Comment (urseberry) [inline README.md:41]:
```suggestion
   in or under the location specified in `plugin.path` . If you use Confluent Platform, run
```

File: README.md [hunk 6]
```diff
@@ -107,6 +107,7 @@ A REST call can be executed against one of the cluster instances, and the config
 | `datadog.retry.max` | The number of retries before the output plugin stops.                                                                                                      | `5` ||
 | `datadog.retry.backoff_ms` | The time in milliseconds to wait following an error before a retry attempt is made.                                                                        | `3000` ||
 | `datadog.add_published_date` | Valid settings are true or false. When set to `true`, The timestamp is retrieved from the Kafka record and passed to Datadog as `published_date`           ||
+| `datadog.parse_record_headers` | Valid settings are true or false. When set to `true`, Kafka Record Headers will be parsed and passed to DataDog as `kafkaheaders` object |`false`|
```
Comment (urseberry) [inline README.md:110]:
```suggestion
| `datadog.parse_record_headers` | Valid settings are true or false. When set to `true`, Kafka Record Headers are parsed and passed to DataDog as a `kafkaheaders` object |`false`| 
```

File: README.md [hunk 7]
```diff
@@ -135,7 +136,7 @@ transforms.addExtraField.type=org.apache.kafka.connect.transforms.InsertField$Va
 transforms.addExtraField.static.field=extraField
 transforms.addExtraField.static.value=extraValue
 ```
-Now if you restart the sink connector and send some more test messages, each new record should have a `extraField` field 
+Now if you restart the sink connector and send some more test messages, each new record should have a `extraField` field
```
Comment (urseberry) [inline README.md:139]:
```suggestion
If you restart the sink connector and send some more test messages, each new record should have a `extraField` field
```

File: README.md [hunk 8]
```diff
@@ -146,14 +147,14 @@ To run the supplied unit tests, run `mvn test` from the root of the project.
 
 ### System Tests
 
-We use Confluent Platform for a batteries-included Kafka environment for local testing. Follow the guide 
+We use Confluent Platform for a batteries-included Kafka environment for local testing. Follow the guide
```
Comment (urseberry) [inline README.md:150]:
```suggestion
Use Confluent Platform for a batteries-included Kafka environment for local testing. Follow the guide
```

File: README.md [hunk 9]
```diff
@@ -146,14 +147,14 @@ To run the supplied unit tests, run `mvn test` from the root of the project.
...
 [here](https://docs.confluent.io/current/quickstart/ce-quickstart.html) to install the Confluent Platform.
 
-Then, install the [Confluent Kafka Datagen Connector](https://github.com/confluentinc/kafka-connect-datagen) to create 
-sample data of arbitrary types. Install this Datadog Logs Connector by running 
+Then, install the [Confluent Kafka Datagen Connector](https://github.com/confluentinc/kafka-connect-datagen) to create
+sample data of arbitrary types. Install this Datadog Logs Connector by running
 `confluent-hub install target/components/packages/<connector-zip-file>`.
 
-In the `/test` directory there are some `.json` configuration files to make it easy to create Connectors. There are 
+In the `/test` directory there are some `.json` configuration files to make it easy to create Connectors. There are
```
Comment (urseberry) [inline README.md:157]:
```suggestion
In the `/test` directory, there are some `.json` configuration files to make it easy to create Connectors. There are
```

File: README.md (hunk 4 above)
Comment (jm1221d) [reply README.md]: Fixed.

File: README.md (hunk 5 above)
Comment (jm1221d) [reply README.md]: Fixed.

File: README.md (hunk 6 above)
Comment (jm1221d) [reply README.md]: Fixed.

File: README.md (hunk 7 above)
Comment (jm1221d) [reply README.md]: Fixed.

File: README.md (hunk 8 above)
Comment (jm1221d) [reply README.md]: Fixed.

File: README.md (hunk 9 above)
Comment (jm1221d) [reply README.md]: Fixed.

File: src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java (hunk 2 above)
Comment (jm1221d) [reply src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java]: Reverted.

File: src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java (hunk 1 above)
Comment (jm1221d) [reply src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java]: Hi @remeh, I have a question regarding the more precise serialization. I'm not so familiar with Gson, so I wanted to ask you if there is a more efficient way of doing the above conversion from `Header` to `JsonElement`. One approach that I found was to return a `JsonObject` and use toJsonTree method, but then we would serialize `N` times, where `N` is the number of `Header` objects. The above solution would have a constant (2) number of calls to the Gson library, no matter how many `Header` objects we have.

**Review by urseberry — APPROVED** [review]

File: src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java (hunk 1 above)
Comment (remeh) [reply src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java]: > We could add an additional property that would contain an explicit list of headers that the user would like to inject into DataDog.

I'm not very aware of the general use-case of this PR but I think this addition could be interesting to have in the first iteration, yes. But then a new question enters the scene: what if the additional property isn't configured? Should we serialise all the headers? I think so if/since the user has explicitly set the `datadog.parse_record_headers` to `true`. Let me know if you have a different opinion.

>  One approach that I found was to return a JsonObject and use toJsonTree method, but then we would serialize N times, where N is the number of Header objects.

I don't think calling `N` times the serialisation is a problem, but it's probably not faster than calling it once for all headers anyway. Maybe you'll want to do the former (per header) when there is a filter and the latter (the whole object) when there is no filter?

Comment (jm1221d) [reply src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java]: > I'm not very aware of the general use-case of this PR but I think this addition could be interesting to have in the first iteration, yes. But then a new question enters the scene: what if the additional property isn't configured? Should we serialise all the headers? I think so if/since the user has explicitly set the `datadog.parse_record_headers` to `true`. Let me know if you have a different opinion.
> 

This PR would solve the missing Kafka Record Headers in the DataDog Logs UI. For example, you publish two identical events into Kafka `{"foo": "foo", "bar": "bar"}`, that, downstream, produce multiple events that cannot be correlated in a deterministic way, and let's say you are not in a position to introduce an attribute in the event payload which would serve as a unique identifier. In that case, you can add a header containing a unique identifier (event payload would remain the same), and that header would be propagated to the downstream services and the downstream event would contain that header. The current DataDog Kafka Connector connector issue is that the headers are not parsed and sent to DataDog -> correlation between messages is almost impossible for the above scenario. 

Regarding the additional attribute, maybe we could go with the above approach for the first iteration. If the additional property is defined, then we would parse those headers that are explicitly stated, and if not we would parse them all.

Regarding the serialization, the current intention would be to have them as a `Map<String, Object>`, so one serialization call would be preferred from my perspective. The `toJson/fromJson` Gson methods were used because those were the only way I knew how to achieve the conversion from `Map<String, Object>` to JsonElement without serializing/deserializing multiple times.

File: src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java (hunk 3 above)
Comment (jm1221d) [reply src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java:118]: kafkaHeadersToJsonElement method is passed as a Supplier into the populateMetadata. That means that the code is executed when .get() is called, in our case here:
`if (config.parseRecordHeaders) {
            content.add("kafkaheaders", kafkaHeaders.get());
        }`

File: src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java (hunk 1 above)
Comment (remeh) [reply src/main/java/com/datadoghq/connect/logs/sink/DatadogLogsApiWriter.java]: It makes sense to me, especially since this is behind a configuration flag and not enabled by default 👍.

**Review by remeh — APPROVED** [review]

**Comment by remeh** [comment]
Thanks for the contribution @jm1221d. I'll try to have a new release v1.3.0 shipped this week.
